bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ReadOnlyLedgerHandle.java

```diff
@@ -326,11 +326,15 @@ CompletableFuture<Versioned<LedgerMetadata>> closeRecovered() {
                     return builder.withClosedState().withLastEntryId(lac).withLength(len).build();
                 },
                 this::setLedgerMetadata).run();
-        f.thenRun(() -> {
-                synchronized (metadataLock) {
-                    newEnsemblesFromRecovery.clear();
-                }
-            });
+        f.whenComplete((result, exception) -> {
+            synchronized (metadataLock) {
+                newEnsemblesFromRecovery.clear();
+            }
+            if (exception != null) {
```

Comment:
The exception has been catched in the outside logic, and where is the memory leak?

Reply:
The promise may be complete with an exception. https://github.com/apache/bookkeeper/blob/01232c94c91759a175345e8c055951e5cc6091dd/bookkeeper-server/src/main/java/org/apache/bookkeeper/client/MetadataUpdateLoop.java#L125-L178

Reply:
yes, MetadataUpdateLoop.run may be completed with an exception. @hangc0276